peachjam/js/components/pdf-renderer.ts

```diff
@@ -37,40 +35,58 @@ class PdfRenderer {
       }
     });
-    observer.observe(this.root, { attributes: true });
-    this.setupPdfAndPreviewPanels().then(() => {
-      this.root.removeAttribute('data-loading-progress');
-      this.root.removeAttribute('data-pdf-loading');
-      this.root.dispatchEvent(new CustomEvent('pdf-loaded'));
-
-      const pages: Array<HTMLElement> = Array.from(this.root.querySelectorAll('.pdf-content__page'));
-      const previewPanels = Array.from(root.querySelectorAll('.preview-panel'));
-      for (const previewPanel of previewPanels) {
-        previewPanel.addEventListener('click', (e) => this.handlePreviewPanelClick(e));
-      }
+    loadingProgressObserver.observe(this.root, {
+      attributeFilter: ['data-loading-progress']
+    });
 
-      window.addEventListener('scroll', debounce(() => {
-        if (!this.scrollListenerActive) return;
-        let current: HTMLElement | null;
-        for (const page of pages) {
-          if (!(window.scrollY >= page.offsetTop)) return;
-          current = root.querySelector(`.preview-panel[data-page="${page.dataset.page}"]`);
-          if (!current) return;
-          this.activatePreviewPanel(current);
-          const scrollableContainer = this.root.querySelector('[data-preview-scroll-container]');
-          if (!scrollableContainer) return;
-          scrollableContainer.scrollTop = (current.offsetTop + current.clientHeight) - (current.offsetHeight * 2);
+    const startPdfSetupObserver = new MutationObserver(() => {
+      if (!Object.keys(this.root.dataset).includes('startPdfSetup')) return;
+      this.setupPdfAndPreviewPanels().then(() => {
+        this.root.removeAttribute('data-loading-progress');
+        this.root.removeAttribute('data-pdf-loading');
+        this.root.dispatchEvent(new CustomEvent('pdf-loaded'));
+
+        const pages: Array<HTMLElement> = Array.from(this.root.querySelectorAll('.pdf-content__page'));
+        const previewPanels = Array.from(root.querySelectorAll('.preview-panel'));
+        for (const previewPanel of previewPanels) {
+          previewPanel.addEventListener('click', (e) => this.handlePreviewPanelClick(e));
         }
-      }, 20));
+
+        window.addEventListener('scroll', debounce(() => {
+          if (!this.scrollListenerActive) return;
+          let current: HTMLElement | null;
+          for (const page of pages) {
+            if (!(window.scrollY >= page.offsetTop)) return;
+            current = root.querySelector(`.preview-panel[data-page="${page.dataset.page}"]`);
+            if (!current) return;
+            this.activatePreviewPanel(current);
+            const scrollableContainer = this.root.querySelector('[data-preview-scroll-container]');
+            if (!scrollableContainer) return;
+            scrollableContainer.scrollTop = (current.offsetTop + current.clientHeight) - (current.offsetHeight * 2);
+          }
+        }, 20));
+      });
     });
+
+    startPdfSetupObserver.observe(this.root, {
```

Comment:
Why use an observer like this? It seems a bit complex and indirect. This component controls everything, why not have a `loadPdf` method that is called either when the button is clicked, or when the component loads (if the PDF isn't too large)?

Reply:
refactored